hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,344 @@
+    2. Read the remote chunks and store them locally and create the local block. Example: 
+       1. `DN 1 -> readChunk(Container #12, Block #3, Chunk #1)  -> DN 2` // Read the chunk from DN 2
+       2. `DN 1 -> readChunk(Container #12, Block #3, Chunk #2)  -> DN 3` // Read the chunk from DN 3
+       3. ... // Continue for all chunks that are in the union of chunks.
+       4. Datanode will validate the checksum of the chunk read from the peer Datanode before using it to reconcile. This is already performed by the existing client code. Thus, no silent data corruption will be introduced.
+2. `DN 1` checks if any blocks are corrupted. For each corrupted Block: 
+    1. Make a union of all chunks and which DN has the chunks. It is possible that certain chunks are corrupted in one of the container replicas. The **union** of chunks represents all the chunks that are present in any of the replicas.
+    2. Read the remote chunks and store them locally. Example: 
+       1. `DN 1 -> readChunk(Container #12, Block #20, Chunk #13)  -> DN 2`
+       2. `DN 1 -> readChunk(Container #12, Block #20, Chunk #21)  -> DN 3`
+       3. ... // Continue for all chunks that are in the union of chunks.
+       4. Datanode will validate the checksum of the chunk read from the peer Datanode before using it to reconcile. This is already performed by the existing client code. Thus, no silent data corruption will be introduced.
+3. `DN 1` deletes any blocks that are marked as deleted.
+    1. The block continues to be in the tree with the updated checksum to avoid redundant Merkle tree updates.
+4. `DN 1` recomputes Merkle tree and sends it to SCM via ICR (Incremental Container Report).
+
+Note: This document does not cover the case where the checksums recorded at the time of write match the chunks locally within a Datanode but differ across replicas. We assume that the replication code path is correct and that the checksums are correct. If this is not the case, the system is already in a failed state and the reconciliation protocol will not be able to recover it. Chunks once written are not updated, thus this scenario is not expected to occur.
+
+#### `getContainerHashes`
+
+When a datanode receives a request to get container hashes. The following steps are performed:
+
+1. If the Merkle Tree is present, return it.
+2. If the Merkle Tree is not present, the call `getContainerHashes` will return and error. 
+   1. Datanode missing the Merkle Tree will be schedule the scanner to calculate the Merkle Tree and then report to SCM.
+3. If container not closed, return error.
+4. If container not found, return error.
+5. The merkle tree returned by this call will represent the status of the data on disk (post scanner scrub).
+
+## Sample scenarios
```

Comment:
This might have discussed somewhere above, but I got the following question:
1. What do we do when only one container available in quasi closed state?
2. When only 2 quasi closed containers are available.
3. When 2 quasi closed containers are available and post reconciliation, if old closed container comes back. ( we discussed this in offline chat, but it would be good to capture those scenarios)
With # 2 like cases, how do we avoid falsely treating container as good? ( if we reconcile from 2 quasi-closed states, then it may not be guaranteed we have all chunks right?) Could we please discuss those scenarios and clarify. Thanks

Reply:
We will dive deeper into the state changes for container post reconciliation in the SCM doc around reconciliation.
You raise a valid point that is beneficial to remember if a container was `QUASI_CLOSED` if it cannot be compared to a `CLOSED` container.